infrastructure

jannick0 opened the pull request:
This commit series covers infrastructure only.  Now the script `matchnames` is a `shell` script wrapper providing the command line option if the extension `aregex` shall be loaded by the corresponding awk script or not. In the former case the function `amatch`, in the latter the Levensthein distance function is used for matches.

Hope this can serve your needs.

jannick0 committed:
matchnames: remove trailing white spaces

jannick0 committed:
matchnames: remove redundant close(OUTFILE) statements

Gawk administers opening and closing output file streams itself. At
at the end of a gawk run it closes all open output streams.  Opening and
closing the same stream might be a hit on the awk script performance.

jannick0 committed:
Makefile: fix generation of .1-documentation files

* Makefile:
  - add a pattern rule %.1:%.md
  - fix prerequisites of PHONY target 'man'
  - force pandoc to generate files with LF as EOL regardless of platform

jannick0 committed:
doc: regenerate .1-documentation files

jannick0 committed:
Makefile: let gawk check to find extension 'aregex' in a portable manner

jannick0 committed:
Makefile: enhance default variable handling for target 'install '

* Makefile:
  - allow variable DESTDIR to be prepended to 'prefix'.
  - fix the installation path for awk scripts to be AWKPATH, defaulting to
    $(PREFIX)/share/awk if AWKPATH is undefined.

jannick0 committed:
Makefile: simplify test whether diff command succeeded

jannick0 committed:
matchnames: add file error handlers

jannick0 committed:
matchnames: switch dynamically between matching function at run-time

The switch effectively depends on whether the function 'amatch' is defined or not.

jannick0 committed:
matchnames: allow to optionally choose the distance function

This commit introduces a shell script (matchnames.sh, to be renamed to
matchnames in the following commit) which functions as a wrapper script to
choose the distance function on the command line by passing '-l aregex'
on to gawk or not.

* Makefile:
  - rename 'matchnames' to 'matchnames.sh'.
* matchnames:
  - add option '-l aregex' (could be '-lregex', too) to the list of
    optional arguments to the awk script wrapper 'matchnames'.
  - fix typo in usage string.
* matchnames.sh: add wrapper script calling awk script 'matchnames'.

jannick0 committed:
matchnames: name wrapper shell script 'matchnames' again

* Makefile:
  - add matchnames.awk to install prerequisites which is to be installed
    as awk script.
* matchnames.awk: renamed from matchnames.
* matchnames: renamed from matchnames.sh.

jannick0 committed:
Makefile: remove redundant target 'aregexversion'

camwebb reviewed (changes requested):
Hi @jannick0,

Some great modification here. Thanks. But...

In `82e35`, `make aregexversion` should fail if the library is not there.  I like using `gawk -l aregex --version` but it needs to create a fail. I changed the version format to `aregex-1.1.0` 

In `182d1`, there is a missing FUZZERR argument for amatch() in do_match() on line 496.

`881c9d` & `9f2a6` & `bb054`: Great idea, and more elegant, but... sorry, I don't want to complicate the set-up with an additional shell script - I want to keep a single Awk file, with the default being not to use aregex() so Windows and Mac users can use the script as is (as long as they have Gawk v4). Also, the arguments to matchnames are already set in many other scripts which I don't want to rewrite to add `-l aregex` Please remove these commits (except for the `match_names` typo in matchnames).

jannick0 commented:
Hi Cam,

after having slept a night over this, things could be more simple I think - which is almost what you request below.  So, we are on the same page. 

> In `82e35`, `make aregexversion` should fail if the library is not there. I like using `gawk -l aregex --version` but it needs to create a fail. I changed the version format to `aregex-1.1.0`

Understood.

Should the Makefile be runnable under Windows' `cmd`  shell or `powershell`, too?  If so which `make` do they have, i.e. can we assume that shell rules are run in is `bash`, or is it required to be completely Windows-compliant (e.g., `rm -f` does not exist on Windows as mentioned somewhere in the code)?  I think it would be more simple if we could assumt to run the Makefile in a *nix-like shell, only.  
 
> In `182d1`, there is a missing FUZZERR argument for amatch() in do_match() on line 496.

Got it.
 
> `881c9d` & `9f2a6` & `bb054`: Great idea, and more elegant, but... sorry, I don't want to complicate the set-up with an additional shell script - I want to keep a single Awk file, with the default being not to use aregex() so Windows and Mac users can use the script as is (as long as they have Gawk v4).

This is exactly where I changed my mind over night and now I am completely with you. 

>  Also, the arguments to matchnames are already set in many other scripts which I don't want to rewrite to add `-l aregex` Please remove these commits (except for the `match_names` typo in matchnames).

No problem.  I'll update things and force-push shortly.

camwebb commented:
> Should the Makefile be runnable under Windows' `cmd` shell or
> `powershell`, too?  If so which `make` do they have, i.e. can we
> assume that shell rules are run in is `bash`, or is it required to be
> completely Windows-compliant (e.g., `rm -f` does not exist on Windows
> as mentioned somewhere in the code)?  I think it would be more simple
> if we could assumt to run the Makefile in a *nix-like shell, only.

No, I think no need to work out how to use `make` on Windows. Please see
[here](https://github.com/camwebb/taxon-tools/#windows) for directions
to run `matchnames` on Windows - only the unified (libraries inside, no
DLL) awk script is needed.
